src/smolvm/runtime/qemu.py

```
@@ -499,53 +500,88 @@ def _control_socket_path(self, vm_id: str) -> Path:
+        """Return backing path, raising when qemu-img cannot inspect the disk."""
         qemu_img = which("qemu-img")
-        if qemu_img is not None:
-            # Check if the source has a backing file
-            info_result = subprocess.run(
-                [str(qemu_img), "info", "--output=json", str(source)],
-                capture_output=True,
-                text=True,
-                check=False,
+        if qemu_img is None:
+            raise SmolVMError("qemu-img is needed to create a full QEMU snapshot.")
+        info = subprocess.run(
```

Ruff (0.15.15)

[error] `subprocess` call: check for execution of untrusted input

(S603)